### PR TITLE
DKG Refactor

### PR DESCRIPTION
**Pull request by gakonst**
- Removes the `examples/` directory and replaces it with dedicated E2E tests under `dkg/node.rs`. Javascript examples have been moved [here](https://github.com/celo-org/blind-threshold-bls-wasm)
- Adds a `BoardPublisher` trait which clients are supposed to be publishing their data to. The previous `board.rs` has been made a test-helper in-memory Board _without any authentication_. As discussed with @kobigurk, we can assume an authenticated channel for this, since in practice we will use an Ethereum smart contract as the board.
- The DKG phases are abstracted in a `DKGPhase` trait with a consistent API across all stages. We implement a state machine which transitions between phases while consuming the previous phase. This ensures that once a phase has been run, you cannot go back (hence relying on the type system for safety)

The "real" board publisher will be `Write` probably such as a File.

### CODE REVIEW

File: src/dkg/board.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+use super::primitives::{BundledJustification, BundledResponses, BundledShares};
+use crate::group::Curve;
+
+/// Trait which must be implemented for writing to the board. This trait assumes
+/// an authenticated channel.
+pub trait BoardPublisher<C: Curve> {
+    type Error;
+
+    /// Publishes the shares to the board
+    fn publish_shares(&mut self, shares: BundledShares<C>) -> Result<(), Self::Error>;
```
Comment (dignifiedquire) [inline src/dkg/board.rs:14]: Note: I won't be able to use this trait unfortunately as I need these methods to be `async`, which is a little unfortunate.

Comment (gakonst) [reply src/dkg/board.rs:14]: Yeah I have that in mind, considering moving this to `async_trait`. If you have any code which would potentially consume this, please share:)

Comment (gakonst) [reply src/dkg/board.rs:14]: On further thought, we could have a `BoardPublisher` and `AsyncBoardPublisher` trait for that functionality.

Comment (dignifiedquire) [reply src/dkg/board.rs:14]: @gakonst sent you an invite to a repo, where I am using this atm

**Comment by gakonst** [comment]
Agreed, talked it over in private and I have some ideas on how it'll look

File: src/ffi.rs [hunk 2]
```diff
@@ -635,7 +633,5 @@ mod tests {
         let de = unsafe { de.assume_init() };
 
         assert_eq!(public_key, unsafe { &*de });
-        destroy_pubkey(de as *mut PublicKey);
```
Comment (kobigurk) [inline src/ffi.rs:638]: I'm not sure about this, I'll open a PR

**Pull request merged**